File: apk/app/src/main/java/com/example/monytix/budgetpilot/BudgetPilotViewModel.kt
package com.example.monytix.budgetpilot

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.monytix.data.BackendApi
import com.example.monytix.data.TransactionCreateRequest
import com.example.monytix.data.BudgetRecommendation
import com.example.monytix.data.CommittedBudget
import com.example.monytix.data.BudgetVariance
import com.example.monytix.data.Supabase
import com.example.monytix.goaltracker.GoalUpdateCache
import io.github.jan.supabase.auth.auth
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.LocalDate
import java.time.format.DateTimeFormatter

data class BudgetDeviation(
    val needs: Double,
    val wants: Double,
    val savings: Double
)

data class AutopilotSuggestion(
    val shiftFrom: String,
    val shiftTo: String,
    val pct: Double,
    val message: String
)

data class BudgetPilotUiState(
    val recommendations: List<BudgetRecommendation> = emptyList(),
    val committedBudget: CommittedBudget? = null,
    val variance: BudgetVariance? = null,
    val deviation: BudgetDeviation? = null,
    val autopilotSuggestion: AutopilotSuggestion? = null,
    val budgetState: com.example.monytix.data.BudgetStateResponse? = null,
    val lastUpdatedAt: String? = null,
    val userEmail: String? = null,
    val selectedMonth: String = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE).substring(0, 7),
    val goalsCount: Int = 0,
    val isLoadingRecommendations: Boolean = false,
    val isLoadingCommitted: Boolean = false,
    val isLoadingVariance: Boolean = false,
    val isLoadingState: Boolean = false,
    val isCommitting: Boolean = false,
    val committingPlanCode: String? = null,
    val isApplyingAdjustment: Boolean = false,
    val isRecalculating: Boolean = false,
    val error: String? = null
)

class BudgetPilotViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(BudgetPilotUiState())
    val uiState: StateFlow<BudgetPilotUiState> = _uiState.asStateFlow()

    private fun getAccessToken(): String? =
        Supabase.client.auth.currentSessionOrNull()?.accessToken

    init {
        loadSession()
        loadData()
        // If BudgetUpdateCache has data from a transaction, we'll refresh when screen is shown
        viewModelScope.launch {
            BudgetUpdateCache.consume()?.let { loadData() }
        }
    }

    fun loadSession() {
        viewModelScope.launch {
            val token = getAccessToken() ?: return@launch
            val result = withContext(Dispatchers.IO) { BackendApi.getSession(token) }
            result.getOrNull()?.let { session ->
                _uiState.update { it.copy(userEmail = session.email) }
            }
        }
    }

    fun setMonth(month: String) {
        _uiState.update { it.copy(selectedMonth = month) }
        loadData()
    }

    fun loadData() {
        viewModelScope.launch {
            val token = getAccessToken() ?: return@launch
            val month = _uiState.value.selectedMonth
            val monthParam = if (month.isNotBlank()) month else null

            _uiState.update {
                it.copy(
                    isLoadingState = true,
                    isLoadingRecommendations = true,
                    isLoadingCommitted = true,
                    isLoadingVariance = true,
                    error = null
                )
            }

            val stateResult = withContext(Dispatchers.IO) { BackendApi.getBudgetState(token, monthParam) }
            val goalsResult = withContext(Dispatchers.IO) { BackendApi.getUserGoals(token) }
            val goalsCount = goalsResult.getOrNull()?.count { it.status.lowercase() == "active" } ?: 0

            stateResult.fold(
                onSuccess = { state ->
                    val committedPlan = state.committed_plan
                    val committedBudget = committedPlan?.let { cp ->
                        CommittedBudget(
                            plan_code = cp.plan_id,
                            alloc_needs_pct = (cp.target["needs"] ?: 0.0) / 100.0,
                            alloc_wants_pct = (cp.target["wants"] ?: 0.0) / 100.0,
                            alloc_assets_pct = (cp.target["savings"] ?: 0.0) / 100.0,
                            goal_allocations = emptyList()
                        )
                    }
                    val actual = state.actual
                    val income = state.income_amt
                    val variance = when {
                        committedPlan != null && income > 0 -> {
                            val targetNeeds = (committedPlan.target["needs"] ?: 0.0) / 100.0 * income
                            val targetWants = (committedPlan.target["wants"] ?: 0.0) / 100.0 * income
                            val targetSavings = (committedPlan.target["savings"] ?: 0.0) / 100.0 * income
                            BudgetVariance(
                                income_amt = income,
                                needs_amt = actual?.needs_amt ?: 0.0,
                                planned_needs_amt = targetNeeds,
                                variance_needs_amt = (actual?.needs_amt ?: 0.0) - targetNeeds,
                                wants_amt = actual?.wants_amt ?: 0.0,
                                planned_wants_amt = targetWants,
                                variance_wants_amt = (actual?.wants_amt ?: 0.0) - targetWants,
                                assets_amt = actual?.savings_amt ?: 0.0,
                                planned_assets_amt = targetSavings,
                                variance_assets_amt = (actual?.savings_amt ?: 0.0) - targetSavings
                            )
                        }
                        actual != null && (income > 0 || (actual.needs_amt + actual.wants_amt + actual.savings_amt) > 0) -> {
                            val effectiveIncome = if (income > 0) income else (actual.needs_amt + actual.wants_amt + actual.savings_amt)
                            BudgetVariance(
                                income_amt = effectiveIncome,
                                needs_amt = actual.needs_amt,
                                planned_needs_amt = 0.0,
                                variance_needs_amt = 0.0,
                                wants_amt = actual.wants_amt,
                                planned_wants_amt = 0.0,
                                variance_wants_amt = 0.0,
                                assets_amt = actual.savings_amt,
                                planned_assets_amt = 0.0,
                                variance_assets_amt = 0.0
                            )
                        }
                        else -> null
                    }
                    val deviation = state.deviation?.let { d ->
                        BudgetDeviation(needs = d.needs, wants = d.wants, savings = d.savings)
                    }
                    val suggestion = deviation?.let { d ->
                        when {
                            d.savings < -5 && d.wants > 5 ->
                                AutopilotSuggestion("wants", "savings", minOf(5.0, -d.savings / 2, d.wants), "Move ${minOf(5.0, -d.savings / 2).toInt()}% from wants to savings")
                            d.savings < -5 ->
                                AutopilotSuggestion("wants", "savings", minOf(5.0, -d.savings / 2), "Move ${minOf(5.0, -d.savings / 2).toInt()}% from wants to savings")
                            else -> null
                        }
                    }
                    val recommendations = state.plans.map { p ->
                        BudgetRecommendation(
                            plan_code = p.plan_id,
                            name = p.name,
                            recommendation_reason = p.reason,
                            score = p.score
                        )
                    }
                    _uiState.update {
                        it.copy(
                            budgetState = state,
                            committedBudget = committedBudget,
                            variance = variance,
                            deviation = deviation,
                            autopilotSuggestion = suggestion,
                            recommendations = recommendations,
                            lastUpdatedAt = state.last_updated_at,
                            goalsCount = goalsCount,
                            isLoadingState = false,
                            isLoadingRecommendations = false,
                            isLoadingCommitted = false,
                            isLoadingVariance = false
                        )
                    }
                },
                onFailure = { e ->
                    _uiState.update {
                        it.copy(
                            isLoadingState = false,
                            isLoadingRecommendations = false,
                            isLoadingCommitted = false,
                            isLoadingVariance = false,
                            error = e.message ?: "Failed to load budget state"
                        )
                    }
                }
            )
        }
    }

    fun recalculate() {
        viewModelScope.launch {
            val token = getAccessToken() ?: return@launch
            val month = _uiState.value.selectedMonth
            val monthParam = if (month.isNotBlank()) month else null

            _uiState.update { it.copy(isRecalculating = true) }
            val result = withContext(Dispatchers.IO) {
                BackendApi.recalculateBudget(token, monthParam)
            }
            result.fold(
                onSuccess = { loadData() },
                onFailure = { e -> _uiState.update { it.copy(error = e.message ?: "Recalculate failed") } }
            )
            _uiState.update { it.copy(isRecalculating = false) }
        }
    }

    fun commitBudget(planCode: String) {
        Log.d("BudgetPilot", "commitBudget called: planCode=$planCode")
        viewModelScope.launch {
            val token = getAccessToken()
            if (token == null) {
                Log.e("BudgetPilot", "commitBudget: No access token, user may not be logged in")
                _uiState.update {
                    it.copy(error = "Not logged in. Please sign in again.")
                }
                return@launch
            }
            val month = _uiState.value.selectedMonth
            val monthParam = if (month.isNotBlank()) "${month}-01" else null
            Log.d("BudgetPilot", "commitBudget: calling API planCode=$planCode month=$monthParam")

            _uiState.update {
                it.copy(isCommitting = true, committingPlanCode = planCode)
            }

            val result = withContext(Dispatchers.IO) {
                BackendApi.commitBudget(token, planCode, monthParam)
            }

            Log.d("BudgetPilot", "commitBudget: result isSuccess=${result.isSuccess}")
            result.fold(
                onSuccess = { budget ->
                    Log.d("BudgetPilot", "commitBudget: success plan_code=${budget.plan_code}")
                    _uiState.update {
                        it.copy(
                            committedBudget = budget,
                            isCommitting = false,
                            committingPlanCode = null
                        )
                    }
                    loadData()
                },
                onFailure = { e ->
                    Log.e("BudgetPilot", "commitBudget: failed", e)
                    _uiState.update {
                        it.copy(
                            isCommitting = false,
                            committingPlanCode = null,
                            error = e.message ?: "Failed to commit budget"
                        )
                    }
                }
            )
        }
    }

    fun refresh() {
        BudgetUpdateCache.consume() // clear any stale cache
        loadData()
    }

    fun applyAdjustment(shiftFrom: String, shiftTo: String, pct: Double) {
        viewModelScope.launch {
            val token = getAccessToken() ?: return@launch
            val month = _uiState.value.selectedMonth
            val monthParam = if (month.isNotBlank()) month else null

            _uiState.update { it.copy(isApplyingAdjustment = true) }

            val result = withContext(Dispatchers.IO) {
                BackendApi.applyBudgetAdjustment(token, shiftFrom, shiftTo, pct, monthParam)
            }

            result.fold(
                onSuccess = { resp ->
                    when (resp.status) {
                        "applied" -> loadData()
                        "rejected" -> _uiState.update {
                            it.copy(error = resp.reason ?: "Adjustment rejected")
                        }
                        else -> _uiState.update { it.copy(error = "Unknown response") }
                    }
                    _uiState.update { it.copy(isApplyingAdjustment = false) }
                },
                onFailure = { e ->
                    _uiState.update {
                        it.copy(
                            isApplyingAdjustment = false,
                            error = e.message ?: "Failed to apply adjustment"
                        )
                    }
                }
            )
        }
    }

    fun createTransaction(
        txnDate: String,
        merchantName: String,
        description: String?,
        amount: Double,
        direction: String,
        categoryCode: String?,
        subcategoryCode: String?,
        channel: String?
    ) {
        viewModelScope.launch {
            val token = getAccessToken() ?: return@launch
            _uiState.update { it.copy(isLoadingState = true, error = null) }
            val result = withContext(Dispatchers.IO) {
                BackendApi.createTransaction(
                    token,
                    com.example.monytix.data.TransactionCreateRequest(
                        txn_date = txnDate,
                        merchant_name = merchantName,
                        description = description,
                        amount = amount,
                        direction = direction,
                        category_code = categoryCode,
                        subcategory_code = subcategoryCode,
                        channel = channel
                    )
                )
            }
            result.fold(
                onSuccess = { resp ->
                    BudgetUpdateCache.setFromTransactionCreate(resp.budget_state)
                    GoalUpdateCache.setFromTransactionCreate(resp.updated_goals)
                    _uiState.update { it.copy(isLoadingState = false) }
                    loadData()
                },
                onFailure = { e ->
                    _uiState.update {
                        it.copy(
                            isLoadingState = false,
                            error = e.message ?: "Failed to add transaction"
                        )
                    }
                }
            )
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    private fun computeDeviationAndSuggestion(
        v: BudgetVariance,
        committed: CommittedBudget?
    ): Pair<BudgetDeviation?, AutopilotSuggestion?> {
        if (v.income_amt <= 0 || committed == null) return null to null
        val income = v.income_amt
        val actualNeedsPct = (v.needs_amt / income) * 100
        val actualWantsPct = (v.wants_amt / income) * 100
        val actualSavingsPct = (v.assets_amt / income) * 100
        val plannedNeedsPct = committed.alloc_needs_pct * 100
        val plannedWantsPct = committed.alloc_wants_pct * 100
        val plannedSavingsPct = committed.alloc_assets_pct * 100
        val deviation = BudgetDeviation(
            needs = actualNeedsPct - plannedNeedsPct,
            wants = actualWantsPct - plannedWantsPct,
            savings = actualSavingsPct - plannedSavingsPct
        )
        val suggestion = when {
            deviation.savings < -5 && deviation.wants > 5 ->
                AutopilotSuggestion("wants", "savings", minOf(5.0, -deviation.savings / 2, deviation.wants), "Move ${minOf(5.0, -deviation.savings / 2).toInt()}% from wants to savings")
            deviation.savings < -5 ->
                AutopilotSuggestion("wants", "savings", minOf(5.0, -deviation.savings / 2), "Move ${minOf(5.0, -deviation.savings / 2).toInt()}% from wants to savings")
            else -> null
        }
        return deviation to suggestion
    }
}

File: ios_monytix/ios_monytix/BudgetPilot/BudgetPilotViewModel.swift
//
//  BudgetPilotViewModel.swift
//  ios_monytix
//
//  Mirrors APK BudgetPilotViewModel: budget state, recommendations, commit, apply adjustment, recalculate.
//

import Combine
import Foundation
import SwiftUI

struct GoalProgressForBudget: Identifiable {
    let id: String
    let goalName: String
    let monthlyRequired: Double
}

struct AutopilotSuggestion {
    let shiftFrom: String
    let shiftTo: String
    let pct: Double
    let message: String
}

struct BudgetDeviation {
    let needs: Double
    let wants: Double
    let savings: Double
}

final class BudgetPilotViewModel: ObservableObject {
    @Published private(set) var recommendations: [BudgetRecommendation] = []
    @Published private(set) var committedBudget: CommittedBudget?
    @Published private(set) var variance: BudgetVariance?
    @Published private(set) var deviation: BudgetDeviation?
    @Published private(set) var autopilotSuggestion: AutopilotSuggestion?
    @Published private(set) var budgetState: BudgetStateResponse?
    @Published private(set) var lastUpdatedAt: String?
    @Published var selectedMonth: String = ""
    @Published private(set) var goalsCount: Int = 0
    @Published private(set) var goalsProgressList: [GoalProgressForBudget] = []
    @Published private(set) var goalRequiredMonthlyTotal: Double = 0
    @Published private(set) var goalRequiredSavingsRate: Double = 0
    @Published private(set) var autopilotStatus: String = "Waiting for data"
    @Published private(set) var adaptivePlanReason: [String] = []
    @Published private(set) var userEmail: String?
    @Published private(set) var isLoadingState = false
    @Published private(set) var isCommitting = false
    @Published private(set) var committingPlanCode: String?
    @Published private(set) var isApplyingAdjustment = false
    @Published private(set) var isRecalculating = false
    @Published var errorMessage: String?

    static func currentMonthString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM"
        return formatter.string(from: Date())
    }

    init() {
        selectedMonth = BudgetPilotViewModel.currentMonthString()
        loadSession()
        loadData()
    }

    func loadSession() {
        Task { @MainActor in
            guard let token = await AuthManager.shared.getIdToken() else { return }
            let result = await BackendApi.getSession(accessToken: token)
            if case .success(let session) = result {
                userEmail = session.email
            }
        }
    }

    func setMonth(_ month: String) {
        selectedMonth = month
        loadData()
    }

    func loadData() {
        Task { @MainActor in
            guard let token = await AuthManager.shared.getIdToken() else { return }
            let monthParam = selectedMonth.isEmpty ? nil : selectedMonth
            isLoadingState = true
            errorMessage = nil
            defer { isLoadingState = false }

            async let stateResult = BackendApi.getBudgetState(accessToken: token, month: monthParam)
            async let goalsProgressResult = BackendApi.getGoalsProgress(accessToken: token)
            let (stateRes, goalsRes) = await (stateResult, goalsProgressResult)

            let goalsProgress = (try? goalsRes.get())?.goals ?? []
            goalsCount = goalsProgress.count
            let goalRequiredTotal = goalsProgress.compactMap { $0.monthlyRequired }.reduce(0, +)
            goalRequiredMonthlyTotal = goalRequiredTotal
            goalsProgressList = goalsProgress.compactMap { g in
                guard let req = g.monthlyRequired, req > 0 else { return nil }
                return GoalProgressForBudget(id: g.goalId, goalName: g.goalName, monthlyRequired: req)
            }

            switch stateRes {
            case .success(let state):
                applyBudgetStateSuccess(state, goalRequiredTotal: goalRequiredTotal)
            case .failure(let err):
                errorMessage = BackendApiErrorMessage(err)
            }
        }
    }

    private func applyBudgetStateSuccess(_ state: BudgetStateResponse, goalRequiredTotal: Double) {
        budgetState = state
        lastUpdatedAt = state.lastUpdatedAt

        let committedPlan = state.committedPlan
        if let cp = committedPlan, let target = cp.target {
            let needs = (target["needs"] ?? 0) / 100.0
            let wants = (target["wants"] ?? 0) / 100.0
            let savings = (target["savings"] ?? 0) / 100.0
            committedBudget = CommittedBudget(
                userId: nil,
                month: state.month,
                planCode: cp.planId,
                allocNeedsPct: needs,
                allocWantsPct: wants,
                allocAssetsPct: savings,
                notes: nil,
                committedAt: nil,
                goalAllocations: nil
            )
        } else {
            committedBudget = nil
        }

        let income = state.incomeAmt ?? 0
        let actual = state.actual
        variance = computeVariance(state: state, committedPlan: committedPlan, income: income, actual: actual)

        if let d = state.deviation {
            deviation = BudgetDeviation(
                needs: d.needs ?? 0,
                wants: d.wants ?? 0,
                savings: d.savings ?? 0
            )
        } else {
            deviation = nil
        }

        autopilotSuggestion = computeAutopilotSuggestion(deviation: deviation)

        recommendations = (state.plans ?? []).map { p in
            BudgetRecommendation(
                planCode: p.planId,
                name: p.name,
                description: nil,
                needsBudgetPct: nil,
                wantsBudgetPct: nil,
                savingsBudgetPct: nil,
                score: p.score,
                recommendationReason: p.reason
            )
        }

        let incomeAmt = state.incomeAmt ?? 0
        let savingsAmt = actual?.savingsAmt ?? 0
        let totalSpend = (actual?.needsAmt ?? 0) + (actual?.wantsAmt ?? 0) + savingsAmt
        autopilotStatus = computeAutopilotStatus(
            incomeAmt: incomeAmt,
            totalSpend: totalSpend,
            hasSuggestion: autopilotSuggestion != nil,
            deviation: deviation
        )

        adaptivePlanReason = buildAdaptivePlanReason(
            deviation: deviation,
            goalImpact: state.goalImpact ?? [],
            topPlanReason: recommendations.first?.recommendationReason,
            actual: actual,
            income: incomeAmt
        )
    }

    private func computeVariance(
        state: BudgetStateResponse,
        committedPlan: BudgetStateCommittedPlan?,
        income: Double,
        actual: BudgetStateActual?
    ) -> BudgetVariance? {
        if let cp = committedPlan, let target = cp.target, income > 0 {
            let targetNeeds = (target["needs"] ?? 0) / 100.0 * income
            let targetWants = (target["wants"] ?? 0) / 100.0 * income
            let targetSavings = (target["savings"] ?? 0) / 100.0 * income
            let needsAmt = actual?.needsAmt ?? 0
            let wantsAmt = actual?.wantsAmt ?? 0
            let savingsAmt = actual?.savingsAmt ?? 0
            return BudgetVariance(
                userId: nil,
                month: state.month,
                incomeAmt: income,
                needsAmt: needsAmt,
                plannedNeedsAmt: targetNeeds,
                varianceNeedsAmt: needsAmt - targetNeeds,
                wantsAmt: wantsAmt,
                plannedWantsAmt: targetWants,
                varianceWantsAmt: wantsAmt - targetWants,
                assetsAmt: savingsAmt,
                plannedAssetsAmt: targetSavings,
                varianceAssetsAmt: savingsAmt - targetSavings,
                computedAt: nil
            )
        }
        guard let a = actual else { return nil }
        let total = (a.needsAmt ?? 0) + (a.wantsAmt ?? 0) + (a.savingsAmt ?? 0)
        guard income > 0 || total > 0 else { return nil }
        let effectiveIncome = income > 0 ? income : total
        let needsAmt = a.needsAmt ?? 0
        let wantsAmt = a.wantsAmt ?? 0
        let savingsAmt = a.savingsAmt ?? 0
        return BudgetVariance(
            userId: nil,
            month: state.month,
            incomeAmt: effectiveIncome,
            needsAmt: needsAmt,
            plannedNeedsAmt: 0,
            varianceNeedsAmt: 0,
            wantsAmt: wantsAmt,
            plannedWantsAmt: 0,
            varianceWantsAmt: 0,
            assetsAmt: savingsAmt,
            plannedAssetsAmt: 0,
            varianceAssetsAmt: 0,
            computedAt: nil
        )
    }

    private func computeAutopilotSuggestion(deviation: BudgetDeviation?) -> AutopilotSuggestion? {
        guard let d = deviation else { return nil }
        if d.savings < -5 && d.wants > 5 {
            let pct = min(5.0, min(-d.savings / 2, d.wants))
            return AutopilotSuggestion(
                shiftFrom: "wants",
                shiftTo: "savings",
                pct: pct,
                message: "Move \(Int(pct))% from wants to savings"
            )
        }
        if d.savings < -5 {
            let pct = min(5.0, -d.savings / 2)
            return AutopilotSuggestion(
                shiftFrom: "wants",
                shiftTo: "savings",
                pct: pct,
                message: "Move \(Int(pct))% from wants to savings"
            )
        }
        return nil
    }

    private func computeAutopilotStatus(
        incomeAmt: Double,
        totalSpend: Double,
        hasSuggestion: Bool,
        deviation: BudgetDeviation?
    ) -> String {
        if incomeAmt < 100 && totalSpend < 100 { return "Waiting for data" }
        if hasSuggestion { return "Adjusting" }
        if let d = deviation, (d.savings < -5 || d.needs > 5 || d.wants > 5) { return "Adjusting" }
        return "On Track"
    }

    private func buildAdaptivePlanReason(
        deviation: BudgetDeviation?,
        goalImpact: [BudgetGoalImpact],
        topPlanReason: String?,
        actual: BudgetStateActual?,
        income: Double
    ) -> [String] {
        var reasons: [String] = []
        if let reason = topPlanReason?.trimmingCharacters(in: .whitespacesAndNewlines), !reason.isEmpty {
            reasons.append(reason)
        }
        let totalPlanned = goalImpact.compactMap { $0.plannedAmount }.reduce(0, +)
        if totalPlanned > 0 && income > 0 && (totalPlanned / income) > 0.15 {
            reasons.append("High long-term goal load")
        }
        let needsPct = actual?.needsPct ?? 0
        if needsPct > 55 {
            reasons.append("Above-average EMI commitments")
        }
        if let d = deviation, (abs(d.needs) > 5 || abs(d.wants) > 5 || abs(d.savings) > 5) {
            if !reasons.contains(where: { $0.contains("volatility") }) {
                reasons.append("Moderate income volatility")
            }
        }
        return Array(Set(reasons))
    }

    func recalculate() {
        Task { @MainActor in
            guard let token = await AuthManager.shared.getIdToken() else { return }
            isRecalculating = true
            defer { isRecalculating = false }
            let monthParam = selectedMonth.isEmpty ? nil : selectedMonth
            let result = await BackendApi.recalculateBudget(accessToken: token, month: monthParam)
            switch result {
            case .success:
                loadData()
            case .failure(let err):
                errorMessage = BackendApiErrorMessage(err)
            }
        }
    }

    func commitBudget(planCode: String) {
        Task { @MainActor in
            guard let token = await AuthManager.shared.getIdToken() else {
                errorMessage = "Not signed in. Please sign in again."
                return
            }
            isCommitting = true
            committingPlanCode = planCode
            let result = await BackendApi.commitBudget(accessToken: token, planCode: planCode, month: selectedMonth.isEmpty ? nil : selectedMonth)
            isCommitting = false
            committingPlanCode = nil
            switch result {
            case .success(let budget):
                committedBudget = budget
                loadData()
            case .failure(let err):
                errorMessage = BackendApiErrorMessage(err)
            }
        }
    }

    func refresh() {
        loadData()
    }

    func applyAdjustment(shiftFrom: String, shiftTo: String, pct: Double) {
        Task { @MainActor in
            guard let token = await AuthManager.shared.getIdToken() else { return }
            isApplyingAdjustment = true
            defer { isApplyingAdjustment = false }
            let monthParam = selectedMonth.isEmpty ? nil : selectedMonth
            let result = await BackendApi.applyBudgetAdjustment(
                accessToken: token,
                shiftFrom: shiftFrom,
                shiftTo: shiftTo,
                pct: pct,
                month: monthParam
            )
            switch result {
            case .success(let resp):
                if resp.status == "applied" {
                    loadData()
                } else if resp.status == "rejected" {
                    errorMessage = resp.reason ?? "Adjustment rejected"
                } else {
                    errorMessage = "Unknown response"
                }
            case .failure(let err):
                errorMessage = BackendApiErrorMessage(err)
            }
        }
    }

    func clearError() {
        errorMessage = nil
    }
}

private func BackendApiErrorMessage(_ err: BackendApiError) -> String {
    switch err {
    case .invalidURL:
        return "Invalid URL"
    case .httpStatus(_, let body):
        return body ?? "Request failed"
    case .decoding(let e):
        return "Data error: \(e.localizedDescription)"
    case .network(let e):
        return e.localizedDescription
    }
}
